doc/prebuilt/man/man1/autotune.1
'\" t
.\"     Title: autotune
.\"    Author: [see the "AUTHOR" section]
.\" Generator: DocBook XSL Stylesheets v1.79.1 <http://docbook.sf.net/>
.\"      Date: 2014/08/03
.\"    Manual: Xenomai Manual
.\"    Source: Xenomai 3.0.5
.\"  Language: English
.\"
.TH "AUTOTUNE" "1" "2014/08/03" "Xenomai 3\&.0\&.5" "Xenomai Manual"
.\" -----------------------------------------------------------------
.\" * Define some portability stuff
.\" -----------------------------------------------------------------
.\" ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
.\" http://bugs.debian.org/507673
.\" http://lists.gnu.org/archive/html/groff/2009-02/msg00013.html
.\" ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
.ie \n(.g .ds Aq \(aq
.el       .ds Aq '
.\" -----------------------------------------------------------------
.\" * set default formatting
.\" -----------------------------------------------------------------
.\" disable hyphenation
.nh
.\" disable justification (adjust text to left margin only)
.ad l
.\" -----------------------------------------------------------------
.\" * MAIN CONTENT STARTS HERE *
.\" -----------------------------------------------------------------
.SH "NAME"
autotune \- Calibrate the Xenomai core clock timer
.SH "SYNOPSIS"
.sp
\fBautotune\fR [ options ]
.SH "DESCRIPTION"
.sp
\fBautotune\fR is a utility to determine the best calibration values (aka \fIgravity triplet\fR) for the core clock timer\&.
.sp
The time spent traversing the kernel code from the low\-level Xenomai timer handler until the kernel\-based client handler is invoked, is shorter than the time required to schedule in a kernel thread instead\&. It takes even more time to switch in a user\-space thread, which entails changing the current memory address space, performing potentially time\-consuming MMU\-related operations\&.
.sp
For this reason, Xenomai differentiates timers on the target context they activate, among IRQ(handler), kernel and user threads, anticipating the next timer shot accordingly, so that such context is activated as close as possible to the ideal time\&. This anticipation is called the \fIgravity\fR of the clock serving the timer, which is actually a triplet representing the three possible types of contexts the timer can activate\&.
.sp
Therefore, the gravity is a static adjustment value to account for the basic latency of the target system for responding to timer events, as perceived by the client code waiting for the wake up events\&. Such latency is increased by additional factors, such as:
.sp
.RS 4
.ie n \{\
\h'-04'\(bu\h'+03'\c
.\}
.el \{\
.sp -1
.IP \(bu 2.3
.\}
bus or CPU cache latency,
.RE
.sp
.RS 4
.ie n \{\
\h'-04'\(bu\h'+03'\c
.\}
.el \{\
.sp -1
.IP \(bu 2.3
.\}
delay required to program the timer chip for the next shot,
.RE
.sp
.RS 4
.ie n \{\
\h'-04'\(bu\h'+03'\c
.\}
.el \{\
.sp -1
.IP \(bu 2.3
.\}
code running with interrupts disabled on the CPU to receive the IRQ,
.RE
.sp
.RS 4
.ie n \{\
\h'-04'\(bu\h'+03'\c
.\}
.el \{\
.sp -1
.IP \(bu 2.3
.\}
inter\-processor serialization (\fIspinlocks\fR)\&.
.RE
.sp
\fBautotune\fR runs a series of internal calibration tests for estimating the most appropriate gravity values for its real\-time clock timer, retaining the final values\&.
.if n \{\
.sp
.\}
.RS 4
.it 1 an-trap
.nr an-no-space-flag 1
.nr an-break-flag 1
.br
.ps +1
\fBImportant\fR
.ps -1
.br
.sp
\fBautotune\fR requires the \fBCONFIG_XENO_OPT_AUTOTUNE\fR option to be enabled in the kernel configuration\&.
.sp .5v
.RE
.SH "OPTIONS"
.sp
\fBautotune\fR accepts the following options:
.PP
\fB\-\-irq\fR
.RS 4
Estimate the IRQ gravity value, which is the shortest time the platform needs to deliver an IRQ to a Xenomai interrupt handler in kernel space\&.
.RE
.PP
\fB\-\-kernel\fR
.RS 4
Estimate the kernel gravity value, which is the shortest time the platform needs to deliver an IRQ to a RTDM task running in kernel space\&. This delay includes the context switching time\&.
.RE
.PP
\fB\-\-user\fR
.RS 4
Estimate the user gravity value, which is the shortest time the platform needs to deliver an IRQ to a user\-space task/thread running in a Xenomai application process\&. This delay includes the context switching time\&.
.RE
.PP
\fB\-\-period <ns>\fR
.RS 4
Set the sampling period to the given count of nanoseconds\&. The estimation is performed by measuring the jitter between the ideal time at which a timer tick should be received, and the actual time it is eventually received, for a series of ticks\&. This value expresses the delay between two of these ticks\&. If too short, a lockup might occur\&. A commonly observed result is that the larger the delay, the higher the latency, due to CPU cache effects (i\&.e\&. the real\-time code/data is more likely to get evicted from the cachelines as the non real\-time activity can slip in, treading over a larger address space)\&.
.RE
.PP
\fB\-\-reset\fR
.RS 4
Reset the gravity values to their factory defaults\&. These defaults are statically defined by the Xenomai platform code\&.
.RE
.PP
\fB\-\-noload\fR
.RS 4
Disable load generation while auto\-tuning\&.
\fBautotune\fR
runs a load generator internally in parallel to estimating the latency, in order to eliminate irregular delays which tend to appear on fully idle systems\&. Therefore, keeping the load generation enabled most often leads to a more accurate estimation\&.
.RE
.PP
\fB\-\-verbose[=level]\fR
.RS 4
Set verbosity to the desired level, 1 means almost quiet (default), 2 means fully verbose\&.
.RE
.PP
\fB\-\-help\fR
.RS 4
Display a short help\&.
.RE
.sp
If none of \-\-irq, \-\-kernel or \-\-user is given, an estimation is done for each of them in sequence\&.
.SH "VERSIONS"
.sp
\fBautotune\fR appeared in Xenomai 3\&.0 for the \fICobalt\fR real\-time core\&.
.SH "NOTES"
.sp
The auto\-tuning process may take some time for completing the estimation\&. Although this delay may vary across hardware platforms, running for 30 seconds is common\&.
.sp
Once the gravity values are known for a particular hardware, one may write them to /proc/xenomai/clock/coreclck from some system init script to set up the Xenomai core clock accordingly, instead of running the auto\-tuner after each boot e\&.g:
.sp
.if n \{\
.RS 4
.\}
.nf
    /* change the user gravity to 1728 ns (default) */
# echo 1728 > /proc/xenomai/clock/coreclck
    /* change the IRQ gravity to 129 ns */
# echo 129i > /proc/xenomai/clock/coreclck
    /* change the user and kernel gravities to 1728 and 907 ns resp\&. */
# echo "1728u 907k" > /proc/xenomai/clock/coreclck
.fi
.if n \{\
.RE
.\}
.sp
Alternatively, the gravity values can be statically defined in the kernel configuration of the target kernel:
.sp
.RS 4
.ie n \{\
\h'-04'\(bu\h'+03'\c
.\}
.el \{\
.sp -1
.IP \(bu 2.3
.\}
CONFIG_XENO_OPT_TIMING_SCHEDLAT should be assigned the user gravity value\&.
.RE
.sp
.RS 4
.ie n \{\
\h'-04'\(bu\h'+03'\c
.\}
.el \{\
.sp -1
.IP \(bu 2.3
.\}
CONFIG_XENO_OPT_TIMING_KSCHEDLAT should be assigned the kernel gravity value\&.
.RE
.sp
.RS 4
.ie n \{\
\h'-04'\(bu\h'+03'\c
.\}
.el \{\
.sp -1
.IP \(bu 2.3
.\}
CONFIG_XENO_OPT_TIMING_IRQLAT should be assigned the IRQ gravity value\&.
.RE
.SH "AUTHOR"
.sp
\fBautotune\fR was written by Philippe Gerum <rpm@xenomai\&.org>\&.


ChangeLog
2009-03-23  Philippe Gerum  <rpm@xenomai.org>

	* ksrc/skins/vxworks/semLib.c (sem_destroy_internal) : Add missing
	rescheduling. Fix sem_read_proc().

2009-03-19  Philippe Gerum  <rpm@xenomai.org>

	* include/asm-generic/wrappers.h, src/nucleus/shadow.c: Fix
	generic bits for 2.6.29.

2009-03-18  Philippe Gerum  <rpm@xenomai.org>

	* ksrc/skins/native/pipe.c: Fix streaming mode.

2009-03-18  Jan Kiszka  <jan.kiszka@siemens.com>

	* ksrc/nucleus/shadow.c: Trace syscall exits.

	* patching file include/asm-generic/syscall.h,
	ksrc/nucleus/shadow.c: Add xn_sys_current_info.

	* src/rtdk/rt_print.c: suspend/resume printer thread.

	* ksrc/skins/native/syscall.c, src/skins/native/task.c: Support
	T_WARNSW on task creation.

	* src/skins/native/task.c: Avoid lazy non-TLS rt_task_self.

	* src/skins/native/task.c: Silent migration before self-deletion.

	* src/skins, ksrc/nucleus/shadow.c: Rework initial Linux
	schedparam setting.

	* ksrc/nucleus, include/asm-generic, ksrc/arch/generic: Exclude
	CPUs from Xenomai use.

	* ksrc/nucleus/shadow.c: Cleanup taskexit hook.

	* ksrc/nucleus/shadow.c: Suppress switch warning upon signal
	receipt when the process is being debugged.

	* ksrc/nucleus/sched.c: Allow to change the watchdog timeout via a
	module parameter.

	* ksrc/arch/x86/usercopy_{32, 64}.c: Fix miscompilation of
	__do_strncpy_from_user.

	* src/rtdk, include/rtdk.h: Add assert_context helpers.

2009-03-18  Jan Kiszka  <jan.kiszka@siemens.com>

	* ksrc/skins/posix/syscall.c: Do not return -EFAULT on zero-length
	names. Instead, return -EINVAL when trying to create objects with
	empty names.

2009-03-13  Philippe Gerum  <rpm@xenomai.org>

	* ksrc/skins/vrtx/mx.c: Streamline mutex support.

2009-03-12  Philippe Gerum  <rpm@xenomai.org>

	* ksrc/skins/vrtx/task.c, src/skins/vrtx/task.c: Fix task creation.

2009-03-12  Philippe Gerum  <rpm@xenomai.org>

	* ksrc/nucleus/shadow.c, ksrc/nucleus/synch.c: Introduce debug
	option to detect mutexes held in relaxed sections.

2009-03-07  Philippe Gerum  <rpm@xenomai.org>

	* src/skins/posix/shm.c (__wrap_munmap): Forward to glibc upon EBADF.

	* ksrc/nucleus/Config.in, ksrc/arch/{powerpc, x86}/Config.in: Add
	missing SYS_STACKPOOLSZ switch.

2009-02-26  Philippe Gerum  <rpm@xenomai.org>

	* ksrc/arch/powerpc/patches: Update Adeos support to
	2.6.24.6-powerpc-DENX-2.4-03 (compatible with MPC5121-ADS
	hardware, see ksrc/arch/powerpc/patches/README).

2009-02-25  Jan Kiszka  <jan.kiszka@siemens.com>

	* ksrc/arch/x86/nmi.c: Rework NMI watchdog support.
	
2008-02-24  Stephane Fillod  <stephane.fillod@thomson.net>

	* src/skins/posix/shm.c (__wrap_mmap64): Make real call depend on
	configure detection (HAVE_MMAP64).
	
2009-02-24  Yi Li <liyi.dev@gmail.com>

	* configure.in: Remove -mfdpic switch. Set by default by
	bfin-linux-uclibc-gcc, and may cause error with bfin-uclinux-gcc.

2009-02-24  Philippe Gerum  <rpm@xenomai.org>

	* ksrc/arch/{arm, blackfin, powerpc}/hal.c (rthal_timer_request):
	Port x86 fix of race window.

2009-02-24  Jan Kiszka  <jan.kiszka@siemens.com>

	* ksrc/arch/x86/hal-common.c (rthal_timer_request): Fix race.

	* ksrc/nucleus/shadow.c (gatekeeper_thread): Do not refer to
	uninit sched slots.

2009-02-23  Paul Corner <paul_c@tuxcnc.org>

	* debian/prepare-patch.sh: Fix patch generation rules.

2009-02-23  Philippe Gerum  <rpm@xenomai.org>

	* ksrc/arch/x86/patches: Upgrade to 2.6.27.19-x86-2.2-06,
	2.6.28.7-x86-2.2-06.

2009-02-17  Philippe Gerum  <rpm@xenomai.org>

	* ksrc/arch/x86/patches: Upgrade to 2.6.27.13-x86-2.2-05.

	* ksrc/arch/powerpc/patches: Upgrade to
	2.6.28.5-powerpc-DENX-2.4-04.

2009-02-13  Gilles Chanteperdrix  <gilles.chanteperdrix@xenomai.org>

	* src/skins/posix/shm.c (__wrap_mmap, __wrap_mmap64): Map only
	part of the underlying nucleus heap.

	* ksrc/nucleus/heap.c (xnheap_mmap): Allow partial mapping in
	user-space of a nucleus heap.

	* ksrc/nucleus/sched.c, include/nucleus/sched.h: Introduce
	sched->cpu field member.

2009-02-12  Philippe Gerum  <rpm@xenomai.org>

	* ksrc/nucleus/pod.c (xnpod_start_thread): Run start hook for
	user-space threads as well.

2009-02-11  Philippe Gerum  <rpm@xenomai.org>

	* ksrc/skins/posix/shm.c (ftruncate): Remove GFP_USER hint.

2009-02-07  Philippe Gerum  <rpm@xenomai.org>

	* ksrc/arch/blackfin, include/asm-blackfin: Prepare for 2.6.28.

	* ksrc/arch/blackfin/patches: Upgrade to 2.6.28-1.9-00 (GIT).

	* ksrc/nucleus, include/nucleus: Introduce host tick deferral.

2009-02-02  Philippe Gerum  <rpm@xenomai.org>

	* ksrc/skins/native/syscall.c (rt_queue_write): Fix memory leak.

2009-02-01  Philippe Gerum  <rpm@xenomai.org>

	* ksrc/nucleus/heap, include/asm-generic/system.h, ksrc/skins: Fix
	auto-cleanup for shared heaps.

2009-01-31  Gilles Chanteperdrix  <gilles.chanteperdrix@xenomai.org>

	* ksrc/drivers/testing/switchtest.c,
	  src/testsuite/switchtest/switchtest.c: Add stress mode to
	switchtest.

2009-01-30  Philippe Gerum  <rpm@xenomai.org>

	* ksrc/arch/ia64, include/asm-ia64: Itanium port discontinued.

2009-01-30  Philippe Gerum  <rpm@xenomai.org>

	* ksrc/arch/x86/patches: Upgrade to 2.6.27.13-x86-2.2-04,
	2.6.28.2-x86-2.2-05.

2009-01-27  Gilles Chanteperdrix  <gilles.chanteperdrix@xenomai.org>

	* include/asm-x86/bits/pod_{32, 64}.h (xnarch_leave_root): Fix FPU
	backup clobbering.
	
2009-01-27  Philippe Gerum  <rpm@eclipse.xenomai.org>

	* ksrc/nucleus, include/nucleus, ksrc/skins/posix, include/posix,
	sim, include/asm-sim: Introduce POSIX sporadic scheduling (PSS)
	class.

2009-01-21  Gilles Chanteperdrix  <gilles.chanteperdrix@xenomai.org>

	* ksrc/arch/arm/patches: Upgrade to 2.6.28-arm-1.2-00.
	
2009-01-19  Philippe Gerum  <rpm@xenomai.org>

	* ksrc/nucleus/pipe.c, ksrc/skins/native/pipe.c,
	ksrc/skins/rtai/fifo.c: Refactor callback management, fixing a
	severe latency issue, and a potential memory leak in both
	disconnect/release handlers.

	* ksrc/nucleus/pipe.c (xnpipe_disconnect, xnpipe_release):

utils/autotune/autotune.c
/*
 * Copyright (C) 2014 Philippe Gerum <rpm@xenomai.org>
 *
 * Xenomai is free software; you can redistribute it and/or modify it
 * under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * Xenomai is distributed in the hope that it will be useful, but
 * WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with Xenomai; if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA
 * 02111-1307, USA.
 */
#include <fcntl.h>
#include <unistd.h>
#include <getopt.h>
#include <stdlib.h>
#include <stdio.h>
#include <pthread.h>
#include <limits.h>
#include <time.h>
#include <signal.h>
#include <error.h>
#include <sys/cobalt.h>
#include <rtdm/autotune.h>
#include <xenomai/init.h>

static int tune_irqlat, tune_kernlat, tune_userlat;

static int reset, noload, background;

/*
 * --verbosity_level=0 means fully quiet, =1 means almost quiet.
 */
#define verbose (__base_setup_data.verbosity_level)

static const struct option base_options[] = {
	{
#define irq_opt		0
		.name = "irq",
		.has_arg = no_argument,
		.flag = &tune_irqlat,
		.val = 1
	},
	{
#define kernel_opt	1
		.name = "kernel",
		.has_arg = no_argument,
		.flag = &tune_kernlat,
		.val = 1
	},
	{
#define user_opt	2
		.name = "user",
		.has_arg = no_argument,
		.flag = &tune_userlat,
		.val = 1
	},
	{
#define reset_opt	3
		.name = "reset",
		.has_arg = no_argument,
		.flag = &reset,
		.val = 1
	},
	{
#define noload_opt	4
		.name = "noload",
		.has_arg = no_argument,
		.flag = &noload,
		.val = 1
	},
	{
#define period_opt	5
		.name = "period",
		.has_arg = required_argument,
	},
	{
#define background_opt	6
		.name = "background",
		.has_arg = no_argument,
		.flag = &background,
		.val = 1,
	},
	{ /* Sentinel */ }
};

static void *sampler_thread(void *arg)
{
	int fd = (long)arg, ret, n = 0;
	__u64 timestamp = 0;
	struct timespec now;

	for (;;) {
		ret = ioctl(fd, AUTOTUNE_RTIOC_PULSE, &timestamp);
		if (ret) {
			if (errno != EPIPE)
				error(1, errno, "pulse failed");
			timestamp = 0; /* Next tuning period. */
			n = 0;
		} else {
			n++;
			clock_gettime(CLOCK_MONOTONIC, &now);
			timestamp = (__u64)now.tv_sec * 1000000000 + now.tv_nsec;
		}
	}

	return NULL;
}

static void *load_thread(void *arg)
{
	int fdi, fdo, count = 0, wakelim;
	ssize_t nbytes, ret;
	struct timespec rqt;
	char buf[512];

	fdi = open("/dev/zero", O_RDONLY);
	if (fdi < 0)
		error(1, errno, "/dev/zero");

	fdo = open("/dev/null", O_WRONLY);
	if (fdi < 0)
		error(1, errno, "/dev/null");

	rqt.tv_sec = 0;
	rqt.tv_nsec = CONFIG_XENO_DEFAULT_PERIOD * 2;
	wakelim = 20000000 / rqt.tv_nsec;

	for (;;) {
		clock_nanosleep(CLOCK_MONOTONIC, 0, &rqt, NULL);

		if ((++count % wakelim) == 0) {
			cobalt_thread_relax();
			continue;
		}

		nbytes = read(fdi, buf, sizeof(buf));
		if (nbytes <= 0)
			error(1, EIO, "load streaming");
		if (nbytes > 0) {
			ret = write(fdo, buf, nbytes);
			(void)ret;
		}
	}

	return NULL;
}

static void create_sampler(pthread_t *tid, int fd)
{
	struct sched_param param;
	pthread_attr_t attr;
	int ret;

	pthread_attr_init(&attr);
	pthread_attr_setdetachstate(&attr, PTHREAD_CREATE_JOINABLE);
	pthread_attr_setinheritsched(&attr, PTHREAD_EXPLICIT_SCHED);
	pthread_attr_setschedpolicy(&attr, SCHED_FIFO);
	param.sched_priority = 99;
	pthread_attr_setschedparam(&attr, &param);
	ret = pthread_create(tid, &attr, sampler_thread, (void *)(long)fd);
	if (ret)
		error(1, ret, "sampling thread");

	pthread_attr_destroy(&attr);
	pthread_setname_np(*tid, "sampler");
}

static void create_load(pthread_t *tid)
{
	struct sched_param param;
	pthread_attr_t attr;
	int ret;

	pthread_attr_init(&attr);
	pthread_attr_setdetachstate(&attr, PTHREAD_CREATE_JOINABLE);
	pthread_attr_setinheritsched(&attr, PTHREAD_EXPLICIT_SCHED);
	pthread_attr_setschedpolicy(&attr, SCHED_FIFO);
	param.sched_priority = 1;
	pthread_attr_setschedparam(&attr, &param);
	ret = pthread_create(tid, &attr, load_thread, NULL);
	if (ret)
		error(1, ret, "load thread");

	pthread_attr_destroy(&attr);
	pthread_setname_np(*tid, "loadgen");
}

void application_usage(void)
{
        fprintf(stderr, "usage: %s [options]:\n", get_program_name());
	fprintf(stderr, "--irq				tune for interrupt latency\n");
	fprintf(stderr, "--kernel			tune for kernel scheduling latency\n");
	fprintf(stderr, "--user				tune for user scheduling latency\n");
	fprintf(stderr, "    [ if none of --irq, --kernel and --user is given,\n"
  		        "      tune for all contexts ]\n");
	fprintf(stderr, "--period			set the sampling period\n");
	fprintf(stderr, "--reset 			reset core timer gravity to factory defaults\n");
	fprintf(stderr, "--noload			disable load generation\n");
	fprintf(stderr, "--background 			run in the background\n");
}

static void run_tuner(int fd, unsigned int op, int period, const char *type)
{
	struct autotune_setup setup;
	pthread_t sampler;
	__u32 gravity;
	int ret;

	setup.period = period;
	setup.quiet = verbose > 2 ? 0 : 2 - verbose;
	ret = ioctl(fd, op, &setup);
	if (ret)
		error(1, errno, "setup failed (%s)", type);

	if (verbose) {
		printf("%s gravity... ", type);
		fflush(stdout);
	}

	if (op == AUTOTUNE_RTIOC_USER)
		create_sampler(&sampler, fd);

	ret = ioctl(fd, AUTOTUNE_RTIOC_RUN, &gravity);
	if (ret)
		error(1, errno, "tuning failed (%s)", type);

	if (op == AUTOTUNE_RTIOC_USER)
		pthread_cancel(sampler);

	if (verbose)
		printf("%u ns\n", gravity);
}

int main(int argc, char *const argv[])
{
	int fd, period, ret, c, lindex, tuned = 0;
	pthread_t load_pth;
	cpu_set_t cpu_set;
	time_t start;

	period = CONFIG_XENO_DEFAULT_PERIOD;

	for (;;) {
		c = getopt_long_only(argc, argv, "", base_options, &lindex);
		if (c == EOF)
			break;
		if (c == '?') {
			xenomai_usage();
			return EINVAL;
		}
		if (c > 0)
			continue;

		switch (lindex) {
		case period_opt:
			period = atoi(optarg);
			if (period <= 0)
				error(1, EINVAL, "invalid sampling period (default %d)",
				      CONFIG_XENO_DEFAULT_PERIOD);
			break;
		case noload_opt:
		case background_opt:
			break;
		case irq_opt:
		case kernel_opt:
		case user_opt:
		case reset_opt:
			tuned = 1;
			break;
		default:
			return EINVAL;
		}
	}

	CPU_ZERO(&cpu_set);
	CPU_SET(0, &cpu_set);
	ret = sched_setaffinity(0, sizeof(cpu_set), &cpu_set);
	if (ret)
		error(1, errno, "cannot set CPU affinity");

	if (background) {
		signal(SIGHUP, SIG_IGN);
		ret = daemon(0, 0);
		if (ret)
			error(1, errno, "cannot daemonize");
	}

	fd = open("/dev/rtdm/autotune", O_RDONLY);
	if (fd < 0)
		error(1, errno, "cannot open autotune device");

	if (!tuned)
		tune_irqlat = tune_kernlat = tune_userlat = 1;

	if (reset) {
		ret = ioctl(fd, AUTOTUNE_RTIOC_RESET);
		if (ret)
			error(1, errno, "reset failed");
	}

	if (tune_irqlat || tune_kernlat || tune_userlat) {
		if (!noload)
			create_load(&load_pth);
		if (verbose)
			printf("== auto-tuning started, period=%d ns (may take a while)\n",
			       period);
	} else
		noload = 1;

	time(&start);

	if (tune_irqlat)
		run_tuner(fd, AUTOTUNE_RTIOC_IRQ, period, "irq");

	if (tune_kernlat)
		run_tuner(fd, AUTOTUNE_RTIOC_KERN, period, "kernel");

	if (tune_userlat)
		run_tuner(fd, AUTOTUNE_RTIOC_USER, period, "user");

	if (verbose && (tune_userlat || tune_kernlat || tune_userlat))
		printf("== auto-tuning completed after %ds\n",
		       (int)(time(NULL) - start));

	if (!noload)
		pthread_cancel(load_pth);

	close(fd);

	return 0;
}
